src/descriptor/mod.rs

```diff
@@ -317,15 +317,50 @@ impl<Pk: MiniscriptKey> Descriptor<Pk> {
         }
     }
 
+    /// Computes an upper bound on the difference in weight between a
+    /// non-satisfied `TxIn` and a satisfied `TxIn`.
+    ///
+    /// A non-satisfied `TxIn` contains the `scriptSigLen` varint recording the
+    /// value 0 (which is 4WU) and also the `witnessStackLen` for transactions
+    /// that have at least one witness spend, which will also record the value 0
+    /// (which is 1WU).
+    ///
+    /// Hence, "satisfaction weight" contains the additional varint weight of
+    /// `scriptSigLen` and `witnessStackLen` (which occurs when the varint value
+    /// increases over the 1byte threshold), and also the weights of `scriptSig`
```

Comment:
So what this is wants to say about scriptSigLen and witnessStackLen is:
1. The witnessStackLen is is assumed to be 1 for segwit descriptors but 0  for legacy
2. The scriptSigLen is always by default 1 regardless.
3. This function will tell you the include how much satisfying the inputs will increase these over these assumed defaults.

I wonder if is possible to make this clearer.

I made this issue at https://github.com/rust-bitcoin/rust-bitcoin/issues/1411 where it seems I was mistaken about the definition we were using here. I thought we were doing weight over the weight of `TxIn::default()` but we are doing one less than that for segwit descriptors since we are use 1 as a base witness stack len where as it would be 0 for `TxIn::default()`. What do you think about changing the definition to simply the weight over `TxIn::default()` once it is satisfied (assuming `Txin::weight` existed) i.e. this always holds:

```rust 
for i in 0..transaction.input.len() {
    assert_eq!(descriptor_for_input[i].max_weight_to_satisfy(), transaction.input[i].weight() - Txin::default().weight()_
}
```
in the case where transaction is valid and you used the worst satisfaction path from the corresponding input in `descriptor_for_input`.
This might be a bit easier to explain (assuming `TxIn::weight` is coherent and easy to explain itself!).

Reply:
>What do you think about changing the definition to simply the weight over `TxIn::default()` once it is satisfied (assuming `Txin::weight` existed) i.e. this always holds:

I think this is a good idea. I also tripped over the "starting from 1" logic here. It does make sense, and presumably wallet authors want us to return the smallest possible number that is strictly true (or maybe they want us to return the most predictable and reproducible number, so if this is used for validity checks we won't get inconsistencies like the one that crashed Liquid?), so I let it slide ... but IMO it's a bit surprising and confusing.

Reply:
Maybe the right way to think about this function is that it's used for sanity/validity checking, and we want it to be a simple-to-explain value which tries not to do ad-hoc computations that feel likely to change between version of the library.

When wallets actually want to do fee estimation at spend-time, they presumably have more information and can use the planning API #481 and we can put as many accounting tricks as we want into that.

Reply:
Thanks for the feedback!I think you're right, having a coherent definition of `TxIn` and basing the `max_weight_to_satisfy` calculation on that sounds less error-prone for the end user. I'll open a PR on rust-bitcoin soon :)

Reply:
From my understanding, we aren't changing the implementation of `max_weight_to_satisfy`, but changing how we define things to make it clearer to the end user. We just say, "we assume `TxIn::default()`'s weight always includes 1 byte for `size(witness_len)`".

Reply:
No I am suggesting changing the implementation. If `TxIn` were to have a `weight` function then I think it would have 0 weight for the witness tack length for `TxIn::default`.

Reply:
That's not what the current implementation does if I am not mistaken. It assumes 0 witness stack len for non segwit descriptor and 1 for a segwit descriptor.

> But if we assume at least one input is a segwit spend (which is also the "safer" definition)

Why is this safer though. What guarantee does doing this provide? It doesn't guarantee that if you add all the `max_weight_to_satisfy` of all the inputs together that the sum is an upper bound on the weight contribution of the inputs. You have to know whether one input was segwit to know whether to add the 2WU segwit header. Since you always need to figure out whether there is a segwit input anyway then you might as well use it to add +1 WU for all non-segwit inputs while you're at it. This gives you a correct upper bound on the input weight (if you add the weight of the length byte for the number of inputs).

With assuming that the witness stack length is always 1WU for non-segwit inputs you can still get the correct upper bound so it's also a decent choice. You just have to go and subtract 1WU from all the non-segwit inputs in the case that all of them are non-segwit.

As @apoelstra points out there is no great answer here. We should just try and choose the easiest thing to explain correctly. To me the concept of `TxIn::weight` while not elegant is coherent enough to act as a focal point of this definition. Your explanation involving `Txin::max_weight` seems worse in this respect since what does the max weight of a `TxIn` even mean?

Reply:
> That's not what the current implementation does if I am not mistaken. It assumes 0 witness stack len for non segwit descriptor and 1 for a segwit descriptor.

Yep, correct

> Since you always need to figure out whether there is a segwit input anyway then you might as well use that information to add +1 WU for all non-segwit inputs while you're at it.

(Note that this is not always possible to do: if you're creating a tx spending some P2SH utxos that aren't yours, you can't know before the other party signs if they're legacy or segwit wrapped, and `max_weight_to_satisfy` is often used before signing. But in case you don't you can just assume segwit, and you're fine)

I think we all agree that there's no way to produce the right result without the need for a "second pass" after the whole transaction is made.

Since I believe most users will forget to do that next step (and many may not care to do it at all) I think it's better if we return something that can be used "safely" straight away, even if it means potentially overshooting the fee. So always assuming segwit seems the best choice to me.

More advanced users could then check the tx and subtract -1 * n_inputs if the tx is non-segwit. We can provide an example for that and try to document it the best we can.

Reply:
@LLFourn I'm a little confused by what you mean, so to make sure we are on the same page I will give thoughts to your original comment...

> The witnessStackLen is assumed to be 1 for segwit descriptors but 0 for legacy

From my understanding this is not true. The definition of `max_weight_to_satisfy` is the **additional weight** introduced from satisfying a `TxIn`. For satisfying legacy descriptors, we will never add additional items to the witness stack, so the additional weight of witnessStackLen for legacy descriptors is 0.

This is independent to whether `TxIn::default()` (an unsatisfied txin) contains the `witnessStackLen` field.

Reply:
> @LLFourn I'm a little confused by what you mean, so to make sure we are on the same page I will give thoughts to your original comment...
> 
> > The witnessStackLen is assumed to be 1 for segwit descriptors but 0 for legacy
> 
> From my understanding this is not true. The definition of `max_weight_to_satisfy` is the **additional weight** introduced from satisfying a `TxIn`. For satisfying legacy descriptors, we will never add additional items to the witness stack, so the additional weight of witnessStackLen for legacy descriptors is 0.

Correct. I don't think what you said is in contradiction to what I said.

> But if we assume at least one input is a segwit spend (which is also the "safer" definition), the current implementation can be worded as `<miniscript_thing>.max_weight_to_satisfy() - TxIn::default().max_weight()`, where `TxIn::max_weight` assumes segwit spend.

Ok now I understand what you were saying here and I think we are in the same page. I still think this definition is awkward though (I was at least confused by it!).